deltachat/rpc.go

```diff
@@ -67,48 +79,81 @@ func (self *RpcIO) String() string {
 }
 
 func (self *RpcIO) Start() error {
-	if !self.closed {
-		return fmt.Errorf("Rpc is already running")
+	self.mu.Lock()
+	defer self.mu.Unlock()
+
+	if self.ctx != nil && self.ctx.Err() != nil {
+		return ErrRpcRunning
 	}
-	self.closed = false
-	self.cmd = exec.Command(self.Cmd)
+
+	self.ctx, self.cancel = context.WithCancel(context.Background())
+	self.cmd = exec.CommandContext(self.ctx, self.Cmd)
 	if self.AccountsDir != "" {
 		self.cmd.Env = append(os.Environ(), "DC_ACCOUNTS_PATH="+self.AccountsDir)
 	}
 	self.cmd.Stderr = self.Stderr
 	self.stdin, _ = self.cmd.StdinPipe()
 	stdout, _ := self.cmd.StdoutPipe()
+
+	self.events = make(chan notifyParams, self.EventBuffer)
+	self.accountEvents = make(map[AccountId]chan *Event)
+	options := jrpc2.ClientOptions{OnNotify: self.onNotify}
+	self.client = jrpc2.NewClient(channel.Line(stdout, self.stdin), &options)
+
+	go func() {
+		for {
+			select {
+			case <-self.ctx.Done():
+				return
+			case params := <-self.events:
+				channel := self.getEventChannel(AccountId(params.ContextId))
+
+				var sent bool
+				for !sent {
+					select {
+					case <-self.ctx.Done():
+						return
+					case channel <- params.Event:
+						sent = true
+					case <-time.After(time.Second * 1):
```

Comment:
This isn't strictly necessary bit it helps detecting a deadlock because there are no readers.